File: sdk/python/src/syfthub_sdk/hub.py
"""Hub resource for browsing public endpoints.

This module handles discovery and browsing of public endpoints from other users.
For managing your own endpoints, see the `my_endpoints` module.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from syfthub_sdk._pagination import PageIterator
from syfthub_sdk.models import EndpointPublic

if TYPE_CHECKING:
    from syfthub_sdk._http import HTTPClient


class HubResource:
    """Browse and discover public endpoints from the hub.

    Example usage:
        # Browse all public endpoints
        for endpoint in client.hub.browse():
            print(f"{endpoint.path}: {endpoint.name}")

        # Get trending endpoints
        for endpoint in client.hub.trending(min_stars=10):
            print(f"{endpoint.name} - {endpoint.stars_count} stars")

        # Get a specific endpoint by path
        endpoint = client.hub.get("alice/cool-api")
        print(endpoint.readme)

        # Star an endpoint (requires auth)
        client.hub.star("alice/cool-api")

        # Check if you've starred an endpoint
        if client.hub.is_starred("alice/cool-api"):
            print("You've starred this!")

        # Unstar an endpoint
        client.hub.unstar("alice/cool-api")
    """

    def __init__(self, http: HTTPClient) -> None:
        """Initialize hub resource.

        Args:
            http: HTTP client instance
        """
        self._http = http

    def browse(self, *, page_size: int = 20) -> PageIterator[EndpointPublic]:
        """Browse all public endpoints.

        Args:
            page_size: Number of items per page (default 20)

        Returns:
            PageIterator that lazily fetches endpoints
        """

        def fetch_fn(skip: int, limit: int) -> list[dict[str, Any]]:
            response = self._http.get(
                "/api/v1/endpoints/public",
                params={"skip": skip, "limit": limit},
                include_auth=False,
            )
            return response if isinstance(response, list) else []

        return PageIterator(fetch_fn, EndpointPublic, page_size=page_size)

    def trending(
        self,
        *,
        min_stars: int | None = None,
        page_size: int = 20,
    ) -> PageIterator[EndpointPublic]:
        """Get trending endpoints sorted by stars.

        Args:
            min_stars: Minimum number of stars (optional filter)
            page_size: Number of items per page (default 20)

        Returns:
            PageIterator that lazily fetches endpoints
        """

        def fetch_fn(skip: int, limit: int) -> list[dict[str, Any]]:
            params: dict[str, Any] = {"skip": skip, "limit": limit}
            if min_stars is not None:
                params["min_stars"] = min_stars
            response = self._http.get(
                "/api/v1/endpoints/trending",
                params=params,
                include_auth=False,
            )
            return response if isinstance(response, list) else []

        return PageIterator(fetch_fn, EndpointPublic, page_size=page_size)

    def get(self, path: str) -> EndpointPublic:
        """Get an endpoint by its path (owner/slug format).

        This method searches the public endpoints API to find the endpoint,
        which works reliably across all deployment configurations.

        Args:
            path: Endpoint path in "owner/slug" format (e.g., "alice/cool-api")

        Returns:
            The EndpointPublic

        Raises:
            NotFoundError: If endpoint not found
            ValueError: If path format is invalid
        """
        from syfthub_sdk.exceptions import NotFoundError

        owner, slug = self._parse_path(path)

        # Search public endpoints to find the matching one
        # This approach works because /api/v1/endpoints/public is reliably
        # served by the backend API, unlike /{owner}/{slug} which may be
        # intercepted by frontend routing in some deployments.
        for endpoint in self.browse(page_size=100):
            if endpoint.owner_username == owner and endpoint.slug == slug:
                return endpoint

        raise NotFoundError(
            message=f"Endpoint not found: '{path}'",
            detail=f"No public endpoint found with owner '{owner}' and slug '{slug}'",
        )

    def star(self, path: str) -> None:
        """Star an endpoint.

        Args:
            path: Endpoint path in "owner/slug" format

        Raises:
            AuthenticationError: If not authenticated
            NotFoundError: If endpoint not found
        """
        endpoint_id = self._resolve_endpoint_id(path)
        self._http.post(f"/api/v1/endpoints/{endpoint_id}/star")

    def unstar(self, path: str) -> None:
        """Unstar an endpoint.

        Args:
            path: Endpoint path in "owner/slug" format

        Raises:
            AuthenticationError: If not authenticated
            NotFoundError: If endpoint not found
        """
        endpoint_id = self._resolve_endpoint_id(path)
        self._http.delete(f"/api/v1/endpoints/{endpoint_id}/star")

    def is_starred(self, path: str) -> bool:
        """Check if you have starred an endpoint.

        Args:
            path: Endpoint path in "owner/slug" format

        Returns:
            True if starred, False otherwise

        Raises:
            AuthenticationError: If not authenticated
            NotFoundError: If endpoint not found
        """
        endpoint_id = self._resolve_endpoint_id(path)
        response = self._http.get(f"/api/v1/endpoints/{endpoint_id}/starred")
        data = response if isinstance(response, dict) else {}
        return bool(data.get("starred", False))

    def _parse_path(self, path: str) -> tuple[str, str]:
        """Parse an endpoint path into owner and slug.

        Args:
            path: Path in "owner/slug" format

        Returns:
            Tuple of (owner, slug)

        Raises:
            ValueError: If path format is invalid
        """
        parts = path.strip("/").split("/")
        if len(parts) != 2:
            raise ValueError(
                f"Invalid endpoint path: '{path}'. Expected format: 'owner/slug'"
            )
        return parts[0], parts[1]

    def _resolve_endpoint_id(self, path: str) -> int:
        """Resolve an endpoint path to its ID.

        This requires authentication to get the full endpoint details.
        Uses the user's own endpoints API to find the ID.

        Args:
            path: Endpoint path in "owner/slug" format

        Returns:
            The endpoint ID

        Raises:
            NotFoundError: If endpoint not found
            AuthenticationError: If not authenticated
        """
        from syfthub_sdk.exceptions import NotFoundError

        owner, slug = self._parse_path(path)

        # Search the user's endpoints to find the ID
        # This uses /api/v1/endpoints which returns full details including ID
        response = self._http.get(
            "/api/v1/endpoints",
            params={"limit": 100},
        )
        endpoints = response if isinstance(response, list) else []

        for ep in endpoints:
            if ep.get("slug") == slug:
                endpoint_id = ep.get("id")
                if endpoint_id is not None:
                    return int(endpoint_id)

        # If not found in user's endpoints, the endpoint might belong to another user
        # In this case, we need to search public endpoints and use a different approach
        raise NotFoundError(
            message=f"Could not resolve endpoint ID for '{path}'",
            detail="Endpoint not found or you don't have access to get its ID. "
            "Star/unstar operations require the endpoint ID which is only "
            "available for endpoints you own.",
        )

File: sdk/typescript/src/resources/hub.ts
import type { HTTPClient } from '../http.js';
import type {
  EndpointPublic,
  EndpointSearchResult,
  EndpointSearchResponse,
  SearchOptions,
} from '../models/index.js';
import { PageIterator } from '../pagination.js';

/**
 * Options for browsing endpoints.
 */
export interface BrowseOptions {
  /** Number of items per page (default: 20) */
  pageSize?: number;
}

/**
 * Options for trending endpoints.
 */
export interface TrendingOptions {
  /** Minimum number of stars */
  minStars?: number;
  /** Number of items per page (default: 20) */
  pageSize?: number;
}

/**
 * Hub resource for browsing and discovering public endpoints.
 *
 * For managing your own endpoints, see the MyEndpoints resource.
 *
 * @example
 * // Browse all public endpoints
 * for await (const endpoint of client.hub.browse()) {
 *   console.log(`${endpoint.ownerUsername}/${endpoint.slug}: ${endpoint.name}`);
 * }
 *
 * @example
 * // Get trending endpoints
 * for await (const endpoint of client.hub.trending({ minStars: 10 })) {
 *   console.log(`${endpoint.name} - ${endpoint.starsCount} stars`);
 * }
 *
 * @example
 * // Semantic search for endpoints
 * const results = await client.hub.search('machine learning for images');
 * for (const result of results) {
 *   console.log(`${result.ownerUsername}/${result.slug}: ${result.relevanceScore.toFixed(2)}`);
 * }
 *
 * @example
 * // Get a specific endpoint
 * const endpoint = await client.hub.get('alice/cool-api');
 * console.log(endpoint.readme);
 *
 * @example
 * // Star an endpoint (requires auth)
 * await client.hub.star('alice/cool-api');
 *
 * @example
 * // Check if you've starred an endpoint
 * const starred = await client.hub.isStarred('alice/cool-api');
 */
export class HubResource {
  constructor(private readonly http: HTTPClient) {}

  /**
   * Parse an endpoint path into owner and slug.
   *
   * @param path - Path in "owner/slug" format
   * @returns Tuple of [owner, slug]
   */
  private parsePath(path: string): [string, string] {
    const parts = path.replace(/^\/|\/$/g, '').split('/');
    if (parts.length !== 2 || !parts[0] || !parts[1]) {
      throw new Error(`Invalid endpoint path: '${path}'. Expected format: 'owner/slug'`);
    }
    return [parts[0], parts[1]];
  }

  /**
   * Resolve an endpoint path to its ID.
   *
   * This searches the user's own endpoints to find the ID.
   *
   * @param path - Endpoint path in "owner/slug" format
   * @returns The endpoint ID
   */
  private async resolveEndpointId(path: string): Promise<number> {
    const [, slug] = this.parsePath(path);

    // Search the user's endpoints to find the ID
    // This uses /api/v1/endpoints which returns full details including ID
    const endpoints = await this.http.get<Array<{ id?: number; slug?: string }>>(
      '/api/v1/endpoints',
      { limit: 100 }
    );

    for (const ep of endpoints) {
      if (ep.slug === slug && ep.id !== undefined) {
        return ep.id;
      }
    }

    // Import NotFoundError here to avoid circular dependency
    const { NotFoundError } = await import('../errors.js');
    throw new NotFoundError(
      `Could not resolve endpoint ID for '${path}'. ` +
        'Endpoint not found or you don\'t have access to get its ID.'
    );
  }

  /**
   * Browse all public endpoints.
   *
   * @param options - Pagination options
   * @returns PageIterator that lazily fetches endpoints
   */
  browse(options?: BrowseOptions): PageIterator<EndpointPublic> {
    const pageSize = options?.pageSize ?? 20;

    return new PageIterator<EndpointPublic>(async (skip, limit) => {
      return this.http.get<EndpointPublic[]>(
        '/api/v1/endpoints/public',
        { skip, limit },
        { includeAuth: false }
      );
    }, pageSize);
  }

  /**
   * Get trending endpoints sorted by stars.
   *
   * @param options - Filter and pagination options
   * @returns PageIterator that lazily fetches endpoints
   */
  trending(options?: TrendingOptions): PageIterator<EndpointPublic> {
    const pageSize = options?.pageSize ?? 20;

    return new PageIterator<EndpointPublic>(async (skip, limit) => {
      const params: Record<string, unknown> = { skip, limit };
      if (options?.minStars !== undefined) {
        params['minStars'] = options.minStars;
      }
      return this.http.get<EndpointPublic[]>(
        '/api/v1/endpoints/trending',
        params,
        { includeAuth: false }
      );
    }, pageSize);
  }

  /**
   * Search for endpoints using semantic search.
   *
   * Uses RAG-powered semantic search to find endpoints that match the
   * natural language query. Returns results sorted by relevance score.
   *
   * Note: If RAG is not configured on the server (no OpenAI API key),
   * this method returns an empty array.
   *
   * @param query - Natural language search query (e.g., "machine learning models for image classification")
   * @param options - Search options (topK, type filter, minScore)
   * @returns Promise resolving to array of EndpointSearchResult with relevance scores.
   *          Returns empty array if query is too short (<3 chars) or no matches found.
   *
   * @example
   * // Search for machine learning models
   * const results = await client.hub.search('image classification models');
   * for (const result of results) {
   *   console.log(`${result.ownerUsername}/${result.slug}: ${result.relevanceScore.toFixed(2)}`);
   * }
   *
   * @example
   * // Filter by type and minimum score
   * const dataSources = await client.hub.search('customer data', {
   *   type: EndpointType.DATA_SOURCE,
   *   minScore: 0.5,
   * });
   */
  async search(query: string, options: SearchOptions = {}): Promise<EndpointSearchResult[]> {
    const { topK = 10, type, minScore = 0 } = options;

    // Skip search for very short queries
    if (!query || query.trim().length < 3) {
      return [];
    }

    const body: Record<string, unknown> = {
      query: query.trim(),
      top_k: topK,
    };

    if (type !== undefined) {
      body['type'] = type;
    }

    try {
      const response = await this.http.post<EndpointSearchResponse>(
        '/api/v1/endpoints/search',
        body,
        { includeAuth: false }
      );

      // Filter by minScore and return results
      return (response.results ?? []).filter(
        (result) => result.relevanceScore >= minScore
      );
    } catch {
      // Return empty array on any error (e.g., RAG not configured)
      return [];
    }
  }

  /**
   * Get an endpoint by its path.
   *
   * This method searches the public endpoints API to find the endpoint,
   * which works reliably across all deployment configurations.
   *
   * @param path - Endpoint path in "owner/slug" format (e.g., "alice/cool-api")
   * @returns The EndpointPublic
   * @throws {NotFoundError} If endpoint not found
   */
  async get(path: string): Promise<EndpointPublic> {
    const [owner, slug] = this.parsePath(path);

    // Search public endpoints to find the matching one
    // This approach works because /api/v1/endpoints/public is reliably
    // served by the backend API, unlike /{owner}/{slug} which may be
    // intercepted by frontend routing in some deployments.
    for await (const endpoint of this.browse({ pageSize: 100 })) {
      if (endpoint.ownerUsername === owner && endpoint.slug === slug) {
        return endpoint;
      }
    }

    // Import NotFoundError here to avoid circular dependency
    const { NotFoundError } = await import('../errors.js');
    throw new NotFoundError(
      `Endpoint not found: '${path}'. No public endpoint found with owner '${owner}' and slug '${slug}'.`
    );
  }

  /**
   * Star an endpoint.
   *
   * @param path - Endpoint path in "owner/slug" format
   * @throws {AuthenticationError} If not authenticated
   * @throws {NotFoundError} If endpoint not found
   */
  async star(path: string): Promise<void> {
    const endpointId = await this.resolveEndpointId(path);
    // Use POST method for starring (not PATCH)
    await this.http.post<void>(`/api/v1/endpoints/${endpointId}/star`);
  }

  /**
   * Unstar an endpoint.
   *
   * @param path - Endpoint path in "owner/slug" format
   * @throws {AuthenticationError} If not authenticated
   * @throws {NotFoundError} If endpoint not found
   */
  async unstar(path: string): Promise<void> {
    const endpointId = await this.resolveEndpointId(path);
    // Use DELETE method to /star endpoint (not PATCH to /unstar)
    await this.http.delete<void>(`/api/v1/endpoints/${endpointId}/star`);
  }

  /**
   * Check if you have starred an endpoint.
   *
   * @param path - Endpoint path in "owner/slug" format
   * @returns True if starred, False otherwise
   * @throws {AuthenticationError} If not authenticated
   * @throws {NotFoundError} If endpoint not found
   */
  async isStarred(path: string): Promise<boolean> {
    const endpointId = await this.resolveEndpointId(path);
    const response = await this.http.get<{ starred: boolean }>(
      `/api/v1/endpoints/${endpointId}/starred`
    );
    return response.starred ?? false;
  }
}
